src/Microsoft.AspNetCore.JsonPatch/Internal/DictionaryAdapterOfTU.cs

```diff
@@ -18,10 +18,12 @@ public class ExpandoObjectAdapter : IAdapter
         {
             var contract = (JsonDictionaryContract)contractResolver.ResolveContract(target.GetType());
             var key = contract.DictionaryKeyResolver(segment);
-            var dictionary = (IDictionary<string, object>)target;
+            var dictionary = (IDictionary<TKey, TValue>)target;
 
             // As per JsonPatch spec, if a key already exists, adding should replace the existing value
-            dictionary[key] = ConvertValue(dictionary, key, value);
+            var convertedKey = TryParse<TKey>(key);
+            var convertedValue = TryParse<TValue>(value);
+            dictionary[convertedKey] = ConvertValue(dictionary, convertedKey, convertedValue);
```

Comment:
Instead of TryParse methods, could you use `ConversionResultProvider.ConvertTo` for `TKey` (this could also solve the issue of non-string keys in dictionary) and use `ConvertValue` for `TValue` (you need not use TryParse for value as you are anyway going to convert the value)?